crates/matrix-sdk-sqlite/src/event_cache_store.rs

```diff
@@ -0,0 +1,385 @@
+        let uri = self.encode_key(keys::MEDIA, request.source.unique_key());
+        let format = self.encode_key(keys::MEDIA, request.format.unique_key());
+        let data = self.encode_value(content)?;
+        self.acquire().await?.set_media(uri, format, data).await
+    }
+
+    async fn get_media_content(&self, request: &MediaRequest) -> Result<Option<Vec<u8>>> {
+        let uri = self.encode_key(keys::MEDIA, request.source.unique_key());
+        let format = self.encode_key(keys::MEDIA, request.format.unique_key());
+        let data = self.acquire().await?.get_media(uri, format).await?;
+        data.map(|v| self.decode_value(&v).map(Into::into)).transpose()
+    }
+
+    async fn remove_media_content(&self, request: &MediaRequest) -> Result<()> {
+        let uri = self.encode_key(keys::MEDIA, request.source.unique_key());
+        let format = self.encode_key(keys::MEDIA, request.format.unique_key());
+        self.acquire().await?.remove_media(uri, format).await
+    }
+
+    async fn remove_media_content_for_uri(&self, uri: &ruma::MxcUri) -> Result<()> {
+        let uri = self.encode_key(keys::MEDIA, uri);
+        self.acquire().await?.remove_uri_medias(uri).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        sync::atomic::{AtomicU32, Ordering::SeqCst},
+        time::Duration,
+    };
+
+    use matrix_sdk_base::{
+        event_cache_store::{EventCacheStore, EventCacheStoreError},
+        event_cache_store_integration_tests,
+        media::{MediaFormat, MediaRequest, MediaThumbnailSettings},
+    };
+    use matrix_sdk_test::async_test;
+    use once_cell::sync::Lazy;
+    use ruma::{events::room::MediaSource, media::Method, mxc_uri, uint};
+    use tempfile::{tempdir, TempDir};
+
+    use super::SqliteEventCacheStore;
+    use crate::utils::SqliteObjectExt;
+
+    static TMP_DIR: Lazy<TempDir> = Lazy::new(|| tempdir().unwrap());
+    static NUM: AtomicU32 = AtomicU32::new(0);
+
+    async fn get_event_cache_store() -> Result<SqliteEventCacheStore, EventCacheStoreError> {
+        let name = NUM.fetch_add(1, SeqCst).to_string();
+        let tmpdir_path = TMP_DIR.path().join(name);
+
+        tracing::info!("using event cache store @ {}", tmpdir_path.to_str().unwrap());
+
+        Ok(SqliteEventCacheStore::open(tmpdir_path.to_str().unwrap(), None).await.unwrap())
+    }
+
+    event_cache_store_integration_tests!();
+
+    async fn get_event_cache_store_content_sorted_by_last_access(
+        event_cache_store: &SqliteEventCacheStore,
+    ) -> Vec<Vec<u8>> {
+        let sqlite_db = event_cache_store.acquire().await.expect("accessing sqlite db failed");
+        sqlite_db
+            .prepare("SELECT data FROM media ORDER BY last_access DESC", |mut stmt| {
+                stmt.query(())?.mapped(|row| row.get(0)).collect()
+            })
+            .await
+            .expect("querying media cache content by last access failed")
+    }
+
+    #[async_test]
+    async fn test_last_access() {
```

Comment:
Maybe we can introduce this feature in another PR, because I'd really like us to have it for *all* the `EventCacheStore` implementations, if we decide to go further down that road, and not only for a single backend, otherwise it may show at the use sites of the (abstract) event cache store.

Also I think for simplifying testing purposes and consistency of the backends, the timestamps should be provided by the caller; this will make it trivial to emulate time has spent without relying on `tokio::time` in tests, and will make it possible to implement this feature on the memory backend trivially too, I think.

Reply:
It is planned to be for all the implementations in the next PR. The reasons I introduced it like that for the SQLite store is that:

1. This is a brand new database and I know we'll need this column so it seemed like a waste to init the table without this only to have to write a migration in the next PR.
2. Even if we only add the column without more logic, if a client uses that commit before the next PR is merged, the data will be wrong unless we decide to write a migration to clear the database.

So this is basically a placeholder implementation while we wait for the next PR.

Of course I didn't do this for the memory store because it's not persistent so we can change anything without migration.

Reply:
Sounds good!